[Bug] @df.region merges top-level arguments by kernel local parameter name → distinct region inputs  silently aliased

### Describe the bug

When a @df.region contains multiple @df.kernels, the top-level function interface is built by merging arguments on each kernel's local parameter name, not on the region parameter that the kernel argument is bound to. If two kernels reuse a parameter name (e.g. both name their input x), their two distinct region arguments collapse into a single top-level buffer.




### Reproduction

```python
import allo
from allo.ir.types import int32
import allo.dataflow as df
import numpy as np


@df.region()
def top(a: int32[4], b: int32[4]):             # two DISTINCT region inputs
    @df.kernel(mapping=[1], args=[a])
    def k_a(x: int32[4]):                       # local param name "x"
        for i in range(4):
            x[i] = 1

    @df.kernel(mapping=[1], args=[b])
    def k_b(x: int32[4]):                       # SAME local param name "x" -> bug
        for i in range(4):
            x[i] = 2


if __name__ == "__main__":
    mod = df.build(top, target="simulator")

    a = np.zeros(4, dtype=np.int32)
    b = np.zeros(4, dtype=np.int32)

    # The region has two inputs, so we pass two arrays. The merge bug collapsed
    # them into one top-level argument, so this call FAILS:
    #   AssertionError: # of input arguments mismatch, got 2 but expected 1
    mod(a, b)

    print("a:", a, " b:", b)   # not reached while the bug is present

```
### Buggy output
```txt
assert len(args) == len(
           ^^^^^^^^^^^^^^^^^
AssertionError: # of input arguments mismatch, got 2 but expected 1
```

### Expected behavior

Top-level @top has 2 memref arguments (a, b).


### Additional context

_No response_

--- a/allo/dataflow.py
+++ b/allo/dataflow.py
@@ -462,23 +462,24 @@ def _build_top(s, stream_info, enable_layout=False):
 
     # create argument mapping
     funcs = get_all_df_kernels(s)
-    input_types = []
-    input_signed = ""
     arg_mapping = {}
-    used_args = {}  # {arg_name: arg_idx in top_func}
+    top_args = {}
+    input_types = []
+    for idx, top_arg in enumerate(s.func_args[s.top_func_name]):
+        input_types.append((top_arg.shape, top_arg.dtype))
+        top_args[top_arg.name] = idx
+    input_signed = ["?"] * len(input_types)
     for func in funcs:
         func_name = func.attributes["sym_name"].value
         arg_mapping[func_name] = []
         for i, arg in enumerate(func.arguments):
             if "!allo.stream" not in str(arg.type):
-                arg_name = s.func_args[func_name][i].name
-                if arg_name not in used_args:
-                    used_args[arg_name] = len(input_types)
-                    dtensor = s.func_args[func_name][i]
-                    input_types.append((dtensor.shape, dtensor.dtype))
-                    if "itypes" in func.attributes:
-                        input_signed += func.attributes["itypes"].value[i]
-                arg_mapping[func_name].append(used_args[arg_name])
+                arg_name = s.func_args[func_name][i].top_name
+                if "itypes" in func.attributes:
+                    input_signed[top_args[arg_name]] = func.attributes["itypes"].value[
+                        i
+                    ]
+                arg_mapping[func_name].append(top_args[arg_name])
     # update top function
     top_func = None
     for func in s.module.body.operations:
@@ -497,7 +498,7 @@ def _build_top(s, stream_info, enable_layout=False):
         new_top = func_d.FuncOp(
             name=s.top_func_name, type=func_type, ip=InsertionPoint(top_func)
         )
-        new_top.attributes["itypes"] = StringAttr.get(input_signed)
+        new_top.attributes["itypes"] = StringAttr.get("".join(input_signed))
         new_top.add_entry_block()
         return_op = func_d.ReturnOp([], ip=InsertionPoint(new_top.entry_block))
         for op in top_func.entry_block.operations:

--- a/tests/dataflow/aie/test_matrix.py
+++ b/tests/dataflow/aie/test_matrix.py
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import allo
-from allo.ir.types import int16, int32
+from allo.ir.types import int16, int32, Stream
 import allo.dataflow as df
 import numpy as np
 from allo.memory import Layout
@@ -185,12 +185,57 @@ def test_gemm_2D_mixed():
         print("MLIR_AIE_INSTALL_DIR unset. Skipping AIE backend test.")
 
 
+def test_gemm_atb():
+    Ty = int16
+    M, N, K = 16, 16, 16
+    P0 = 2
+    Ma = M // P0
+
+    @df.region()
+    def top1(A0: Ty[Ma, K], A1: Ty[Ma, K], B: Ty[K, N], C: Ty[M, N]):
+        pipeB: Stream[Ty[K, N], 1][P0]
+        pipeC: Stream[Ty[Ma, N], 1][P0]
+
+        @df.kernel(mapping=[1], args=[B])
+        def loadB(local_B: Ty[K, N]):
+            b = local_B
+            pipeB[0].put(b)
+            pipeB[1].put(b)
+
+        @df.kernel(mapping=[1], args=[A0])
+        def gemm0(local_A: Ty[Ma, K]):
+            c0 = allo.matmul(local_A, pipeB[0].get())
+            pipeC[0].put(c0)
+
+        @df.kernel(mapping=[1], args=[A1])
+        def gemm1(local_A: Ty[Ma, K]):
+            c1 = allo.matmul(local_A, pipeB[1].get())
+            pipeC[1].put(c1)
+
+        @df.kernel(mapping=[1], args=[C])
+        def store(local_C: Ty[M, N]):
+            local_C[:Ma, :] = pipeC[0].get()
+            local_C[Ma:, :] = pipeC[1].get()
+
+    A = np.random.randint(0, 64, (M, K)).astype(np.int16)
+    B = np.random.randint(0, 64, (K, N)).astype(np.int16)
+    C = np.zeros((M, N)).astype(np.int16)
+    if is_available():
+        mod = df.build(top, target="aie")
+        mod(A[:Ma, :], A[Ma:, :], B, C)
+        np.testing.assert_allclose(C, A @ B, atol=1e-5)
+        print("PASSED!")
+    else:
+        print("MLIR_AIE_INSTALL_DIR unset. Skipping AIE backend test.")
+
+
 if __name__ == "__main__":
     test_matrix_scalar_add()
     test_matrix_matrix_add()
     test_gemm_1D_mixed()
     test_gemm_2D_mixed()
     test_gemm_1D()
+    test_gemm_atb()
 
     # Allo flow
     test_gemm_2D()

--- a/tests/dataflow/test_df_unit.py
+++ b/tests/dataflow/test_df_unit.py
@@ -160,8 +160,8 @@ def test_arg_mapping():
 
 
 if __name__ == "__main__":
-    # test_uint()
-    # test_func_index()
-    # test_const_arrays()
-    # test_const_arrays_arithmetic()
+    test_uint()
+    test_func_index()
+    test_const_arrays()
+    test_const_arrays_arithmetic()
     test_arg_mapping()

--- a/tests/dataflow/test_df_unit.py
+++ b/tests/dataflow/test_df_unit.py
@@ -107,8 +107,61 @@ def test_const_arrays_arithmetic():
     print("Dataflow Simulator (Arithmetic) Passed!")
 
 
+def test_arg_mapping():
+    @df.region()
+    def top1(a: int32[4], b: int32[4]):
+        @df.kernel(mapping=[1], args=[a])
+        def k_a(x: int32[4]):  # local param name "x"
+            for i in range(4):
+                x[i] = 1
+
+        @df.kernel(mapping=[1], args=[b])
+        def k_b(x: int32[4]):  # SAME local param name "x"
+            for i in range(4):
+                x[i] = 2
+
+    mod = df.build(top1, target="simulator")
+    a = np.zeros(4).astype(np.int32)
+    b = np.zeros(4).astype(np.int32)
+    mod(a, b)
+    np.testing.assert_allclose(
+        a,
+        np.array([1, 1, 1, 1], dtype=np.int32),
+    )
+    np.testing.assert_allclose(
+        b,
+        np.array([2, 2, 2, 2], dtype=np.int32),
+    )
+
+    @df.region()
+    def top2(a: int32[4], b: int32[4]):
+        @df.kernel(mapping=[1], args=[b])
+        def k_b(x: int32[4]):  # local param name "x"
+            for i in range(4):
+                x[i] = 1
+
+        @df.kernel(mapping=[1], args=[a])
+        def k_a(x: int32[4]):  # SAME local param name "x"
+            for i in range(4):
+                x[i] = 2
+
+    mod = df.build(top2, target="simulator")
+    a = np.zeros(4).astype(np.int32)
+    b = np.zeros(4).astype(np.int32)
+    mod(a, b)
+    np.testing.assert_allclose(
+        a,
+        np.array([2, 2, 2, 2], dtype=np.int32),
+    )
+    np.testing.assert_allclose(
+        b,
+        np.array([1, 1, 1, 1], dtype=np.int32),
+    )
+
+
 if __name__ == "__main__":
-    test_uint()
-    test_func_index()
-    test_const_arrays()
-    test_const_arrays_arithmetic()
+    # test_uint()
+    # test_func_index()
+    # test_const_arrays()
+    # test_const_arrays_arithmetic()
+    test_arg_mapping()

